### PR TITLE
Acceptance tests

### PR DESCRIPTION
**Pull request by bhawanaprasain**
When users are  created in  ownCloud for the first time or   when  previously created users log into  ownCloud  this code does automated testing on activities of Firstrunwizard .

### CODE REVIEW

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+		if ($firstrunwizardScreen != null) {
+			throw new Exception("Popup message from Firstrunwizard is not expected in  second login  ");
+		}
+	}
+
+	/**
+	 * @When the user closes the firstrun wizard popup message
+	 * @return void
+	 */
+
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]: bonus line

File: tests/acceptance/features/webUIFirstrunwizard/firstrunwizard.feature [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  As a user
+  I want to see basic features of ownCloud
+  So that I can get basic overview about ownCloud
+
+  Scenario: first login
+    Given user "user1" has been created with default attributes
+    When user "user1" logs in using the webUI
+    Then the user should see the firstrunwizard popup message
+
+
```
Comment (skshetry) [inline tests/acceptance/features/webUIFirstrunwizard/firstrunwizard.feature]: extra line

File: tests/acceptance/features/webUIFirstrunwizard/firstrunwizard.feature [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  Scenario: second login
+    Given user "user1" has been created with default attributes
+    When user "user1" logs in using the webUI
+    And the user closes the firstrun wizard popup message
+    And the user re-logs in as "user1" using the webUI
+    Then firstrun wizard should not show popup message
```
Comment (dpakach) [inline tests/acceptance/features/webUIFirstrunwizard/firstrunwizard.feature]: Be a bit more consistent in the naming. `firstrun wizard` or `firstrunwizard`?

File: tests/acceptance/features/webUIFirstrunwizard/firstrunwizard.feature [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+Feature:  first login wizard
+  As a user
+  I want to see basic features of ownCloud
+  So that I can get basic overview about ownCloud
+
+  Scenario: first login
+    Given user "user1" has been created with default attributes
+    When user "user1" logs in using the webUI
+    Then the user should see the firstrunwizard popup message
+
```
Comment (dpakach) [inline tests/acceptance/features/webUIFirstrunwizard/firstrunwizard.feature]: delete this extra line

File: tests/acceptance/features/webUIFirstrunwizard/firstrunwizard.feature [hunk 5]
```diff
@@ -0,0 +1,18 @@
+@webUI @insulated @disablePreviews
+Feature:  first login wizard
+  As a user
+  I want to see basic features of ownCloud
```
Comment (dpakach) [inline tests/acceptance/features/webUIFirstrunwizard/firstrunwizard.feature]:
```suggestion
  I want to see basic features of ownCloud when I first log in to owncloud
```

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+	 * @Then the user should see the firstrunwizard popup message
+	 *
+	 * @throws Exception
+	 *
+	 * @return void
+	 */
+	public function theUserShouldSeeTheFirstrunwizardPopupMessage() {
+		$firstrunwizardScreen = $this->firstrunwizardPage->getWizardPopup();
+		if ($firstrunwizardScreen === null) {
+			throw new Exception("could not find wizard");
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]: Better logging would help a lot here. 
```suggestion
			throw new Exception("Could not find firstrunwizard popup.");
```

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+		}
+		if ($firstrunwizardScreen->isVisible() == false) {
+			throw new Exception("CloseWizard  is not visible ");
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]: @bhawanaprasain Is this a CloseWizard? Also, there's two spaces.

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+	 *
+	 * @throws Exception
+	 *
+	 * @return void
+	 */
+	public function theUserShouldNotNotSeeTheFirstrunwizardPopUpMessage() {
+		$firstrunwizardScreen = $this->firstrunwizardPage->getWizardPopup();
+
+		if ($firstrunwizardScreen != null) {
+			throw new Exception("Popup message from Firstrunwizard is not expected in  second login  ");
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]:
```suggestion
			throw new Exception("Popup message from Firstrunwizard is not expected in second login");
```

Comment (skshetry) [reply tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]: Can you rephrase the log message? This specific function/step can be used on different scenarios, not just on second login.

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+		}
+	}
+
+	/**
+	 * @When the user closes the firstrun wizard popup message
+	 * @return void
+	 */
+
+	public function closeDailougeBox() {
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]:
```suggestion
	public function closeDialogueBox() {
```

File: tests/acceptance/features/lib/FirstrunwizardPage.php [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+/**
+ * Page object for Firstrunwizard
+ *
+ */
+
+class FirstrunwizardPage extends OwncloudPage {
+	protected $fileXpath = "//*[@id='closeWizard']";
+
+	protected $secfileXpath = "//*[@id='firstrunwizard']";
```
Comment (skshetry) [inline tests/acceptance/features/lib/FirstrunwizardPage.php]: Can you please rename this to a better variable name?

My suggestion would be:
```suggestion
	protected $firstRunWizardPopupXpath = "//*[@id='firstrunwizard']";
```

File: tests/acceptance/features/lib/FirstrunwizardPage.php [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+	/**
```
Comment (skshetry) [inline tests/acceptance/features/lib/FirstrunwizardPage.php]: Needs additonal line before this line.
```suggestion

	/**
```

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+	/**
+	 * @var FirstrunwizardPage
+	 */
+	private $firstrunwizardPage;
+
+	/**
+	 * WebUIFirstrunwizardContext constructor.
+	 *
+	 */
+	public function __construct(FirstrunwizardPage $firstrunwizardPage) {
```
[the remark this answers is not in the document]
Comment (skshetry) [reply tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]: Will this casing be better? @bhawanaprasain, what do you think?
```suggestion
	public function __construct(FirstRunWizardPage $firstRunWizardPage) {
```

Or, if you don't want to change the file names, just the following:
```suggestion
	public function __construct(FirstrunwizardPage $firstRunWizardPage) {
```

File: tests/acceptance/features/lib/FirstrunwizardPage.php [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+
+use \Behat\Mink\Element\NodeElement;
+
+/**
+ * Page object for Firstrunwizard
+ *
+ */
+
+class FirstrunwizardPage extends OwncloudPage {
+	protected $fileXpath = "//*[@id='closeWizard']";
```
Comment (skshetry) [inline tests/acceptance/features/lib/FirstrunwizardPage.php]: This also needs better variable naming. 
```suggestion
	protected $closeWizardXpath = "//*[@id='closeWizard']";
```

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+use Behat\Behat\Context\Context;
+use Behat\Behat\Hook\Scope\BeforeScenarioScope;
+use Behat\MinkExtension\Context\RawMinkContext;
+use Page\FirstrunwizardPage;
+
+require_once 'bootstrap.php';
+
+/**
+ * Context for password policy specific webUI steps
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]: Need to change documentation:
```suggestion
 * Context for First run wizard
```

**Comment by skshetry** [comment]
Tests are failing as `composer.json` is missing. 
```
+ make test-acceptance-webui
composer install
Do not run Composer as root/super user! See https://getcomposer.org/root for details
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Package operations: 1 install, 0 updates, 0 removals
  - Installing bamarni/composer-bin-plugin (v1.3.0): Downloading (connecting...)Downloading (0%)           Downloading (55%)Downloading (100%)
Generating autoload files
make: *** No rule to make target 'vendor-bin/behat/composer.json', needed by 'vendor-bin/behat/composer.lock'. Stop.
```

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+
+use Behat\Behat\Context\Context;
+use Behat\Behat\Hook\Scope\BeforeScenarioScope;
+use Behat\MinkExtension\Context\RawMinkContext;
+use Page\FirstrunwizardPage;
+
+require_once 'bootstrap.php';
+
+/**
+ * Context for First run wizard  specific webUI steps
```
Comment (dpakach) [inline tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]:
```suggestion
 * Context for First run wizard specific webUI steps
```

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+		if ($firstrunwizardScreen === null) {
+			throw new Exception("Could not find firstrunwizard popup");
+		}
+		if ($firstrunwizardScreen->isVisible() === false) {
+			throw new Exception("Firstrunwizard popup is not visible ");
+		}
+	}
+
+	/**
+	 * @Then firstrunwizard should not show popup message
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]: IMO the scenarios read more nicely if this step is:
```
Then the user should not see the firstrunwizard popup message
```

File: tests/acceptance/features/webUIFirstrunwizard/firstrunwizard.feature [hunk 17]
```diff
@@ -0,0 +1,17 @@
+@webUI @insulated @disablePreviews
+Feature:  first login wizard
+  As a user
+  I want to see basic features of ownCloud when I first log in to owncloud
```
Comment (phil-davis) [inline tests/acceptance/features/webUIFirstrunwizard/firstrunwizard.feature]: Maybe:
```
I want to be informed about the basic features of ownCloud when I first log in
```

File: tests/acceptance/features/webUIFirstrunwizard/firstrunwizard.feature [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+  So that I can get basic overview about ownCloud
```
Comment (phil-davis) [inline tests/acceptance/features/webUIFirstrunwizard/firstrunwizard.feature]:
```
So that I can get a basic overview about ownCloud
```

File: tests/acceptance/features/webUIFirstrunwizard/reset_all.feature [hunk 19]
```diff
@@ -0,0 +1,26 @@
+@webUI @insulated @disablePreviews
+Feature:  first login wizard
```
Comment (phil-davis) [inline tests/acceptance/features/webUIFirstrunwizard/reset_all.feature]: first login wizard is displayed again when the reset-all command has been run

File: tests/acceptance/features/webUIFirstrunwizard/reset_all.feature [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+  As a user
+  I want to see basic features of ownCloud when I first log in to owncloud
+  So that I can get basic overview about ownCloud
```
Comment (phil-davis) [inline tests/acceptance/features/webUIFirstrunwizard/reset_all.feature]: As an administrator
I want to be able to reset the first run wizard display status for all users
So that I can force users to be shown the first run wizard popup again

File: tests/acceptance/features/webUIFirstrunwizard/reset_all.feature [hunk 21]
```diff
@@ -0,0 +1,26 @@
+@webUI @insulated @disablePreviews
```
Comment (phil-davis) [inline tests/acceptance/features/webUIFirstrunwizard/reset_all.feature]: We usually use camel case for the names of feature files. So `resetAll.feature`

File: tests/acceptance/features/webUIFirstrunwizard/reset_all.feature [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+
+  Scenario: Administrator runs reset-all occ command
+    Given user "user1" has been created with default attributes
+    And user "user1" has logged in using the webUI
+    And the user has closed the firstrunwizard popup message
+    And the administrator has invoked occ command "firstrunwizard:reset-all"
+    When the user re-logs in as "user1" using the webUI
+    Then the user should see the firstrunwizard popup message
+
+
```
Comment (phil-davis) [inline tests/acceptance/features/webUIFirstrunwizard/reset_all.feature]: Lots of bonus blank lines here.

Comment (skshetry) [inline tests/acceptance/features/webUIFirstrunwizard/reset_all.feature]: Seems to have accidentally added the file -> empty lines.

File: tests/acceptance/features/webUIFirstrunwizard/reset_all.feature [hunk 23]
```diff
@@ -0,0 +1,26 @@
+@webUI @insulated @disablePreviews
+Feature:  first login wizard
+  As a user
+  I want to see basic features of ownCloud when I first log in to owncloud
+  So that I can get basic overview about ownCloud
+
+  Scenario: Administrator runs reset-all occ command
+    Given user "user1" has been created with default attributes
```
Comment (phil-davis) [inline tests/acceptance/features/webUIFirstrunwizard/reset_all.feature]: We could also create `user2` and not login to `user2` in the given section.
Then also check below that the new `user2` (that is already waiting to see the first run wizard) still sees the popup after the `reset-all` command.
(I can imagine that the `reset-all` command might have a bug where it flips the wizard state for all users, instead of resetting it to "not seen yet")

**Comment by bhawanaprasain** [comment]
@phil-davis please review again

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php (hunk 15 above)
Comment (phil-davis) [reply tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]: Extra space is still there.

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php [hunk 24]
```diff
@@ -0,0 +1,117 @@
...
+	}
+
+	/**
+	 * @When the user closes the firstrunwizard popup message
+	 * @Given the user has closed the firstrunwizard popup message
+
+
+	 * @return void
+	 */
+	public function closeDialougeBox() {
```
[the remark this answers is not in the document]
Comment (phil-davis) [reply tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]: spelling still needs correcting

**Comment by bhawanaprasain** [comment]
@phil-davis please review again

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php [hunk 25]
```diff
@@ -0,0 +1,114 @@
...
+ *
+ */
+
+use Behat\Behat\Context\Context;
+use Behat\Behat\Hook\Scope\BeforeScenarioScope;
+use Behat\MinkExtension\Context\RawMinkContext;
+use Page\FirstrunwizardPage;
+
+require_once 'bootstrap.php';
+/**
```
Comment (dpakach) [inline tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]: Insert one empty line here

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php [hunk 26]
```diff
@@ -0,0 +1,114 @@
...
+	 *
+	 * @throws Exception
+	 *
+	 * @return void
+	 */
+	public function theUserShouldNotNotSeeTheFirstrunwizardPopUpMessage() {
+		$firstrunwizardScreen = $this->firstRunWizardPage->getWizardPopup();
+
+		if ($firstrunwizardScreen !== null) {
+			throw new Exception("Popup message from firstrunwizard is expected only after first login");
```
Comment (dpakach) [inline tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]: Some times we might want to check that the firstrunwizard is not shown when the app is disabled or not installed. In those situations this message may not be appropriate.
How about
```
The firstrunwizard popup is not expected to appear but is present
```
or something like that.

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php [hunk 27]
```diff
@@ -0,0 +1,114 @@
...
+	 * @Then the user should not see the firstrunwizard popup message
+	 *
+	 * @throws Exception
+	 *
+	 * @return void
+	 */
+	public function theUserShouldNotNotSeeTheFirstrunwizardPopUpMessage() {
+		$firstrunwizardScreen = $this->firstRunWizardPage->getWizardPopup();
+
+		if ($firstrunwizardScreen !== null) {
```
Comment (dpakach) [inline tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]: Also sometimes the developers may change the code such that the firstrunwizard is not visible but the element is present in the DOM all times. In such case it may be better to check that if the element is not null than the `$element->isVisible()` is `false`

**Comment by individual-it** [comment]
@phil-davis @skshetry @dpakach please review again

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php [hunk 28]
```diff
@@ -0,0 +1,117 @@
...
+	 *
+	 * @return void
+	 */
+	public function theUserShouldSeeTheFirstrunwizardPopupMessage() {
+		$firstrunwizardScreen = $this->firstRunWizardPage->getWizardPopup();
+		if ($firstrunwizardScreen === null) {
+			throw new Exception("Could not find firstrunwizard popup");
+		}
+		if ($firstrunwizardScreen->isVisible() === false) {
+			throw new Exception("Firstrunwizard popup is not visible ");
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]:
```suggestion
			throw new Exception("Firstrunwizard popup is not visible");
```

**Comment by skshetry** [comment]
Looks good to me. Just a minor suggestion from me.

File: tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php [hunk 29]
```diff
@@ -0,0 +1,117 @@
...
+	 *
+	 * @throws Exception
+	 *
+	 * @return void
+	 */
+	public function theUserShouldNotNotSeeTheFirstrunwizardPopUpMessage() {
+		$firstrunwizardScreen = $this->firstRunWizardPage->getWizardPopup();
+
+		if ($firstrunwizardScreen !== null) {
+			if ($firstrunwizardScreen->isVisible() == true) {
```
Comment (dpakach) [inline tests/acceptance/features/bootstrap/WebUIFirstrunwizardContext.php]:
```suggestion
			if ($firstrunwizardScreen->isVisible() === true) {
```

**Comment by bhawanaprasain** [comment]
@dpakach @skshetry please review again.

File: tests/acceptance/features/bootstrap/bootstrap.php [hunk 30]
```diff
@@ -0,0 +1,34 @@
+<?php
+/**
+ * ownCloud
+ *
+ * @author Artur Neumann <info@jankaritech.com>
```
Comment (dpakach) [inline tests/acceptance/features/bootstrap/bootstrap.php]: I think this name needs to be changed.

Comment (individual-it) [reply tests/acceptance/features/bootstrap/bootstrap.php]: no it does not, I have made the file. It just landed in the same PR as we gave up to split the commits
